src/rust/engine/process_execution/src/remote.rs

```diff
@@ -1721,6 +1822,80 @@ pub mod tests {
       .expect_err("Timeout did not cause failure.");
     assert_contains(&error_msg, "Exceeded time out");
     assert_contains(&error_msg, "echo-a-foo");
+
+    let cancels = mock_server.mock_responder.cancelation_req.lock()
+        .iter()
+        .map(|req| req.get_name().to_owned())
+        .collect::<Vec<_>>();
+    assert_eq!(vec![op_name.to_owned()], cancels);
+  }
+
+  #[test]
+  fn dropped_request_cancels() {
+    let request_timeout = Duration::new(10, 0);
+    let delayed_operation_time = Duration::new(5, 0);
+
+    let execute_request = ExecuteProcessRequest {
+      argv: owned_string_vec(&["/bin/echo", "-n", "foo"]),
+      env: BTreeMap::new(),
+      input_files: EMPTY_DIGEST,
+      output_files: BTreeSet::new(),
+      output_directories: BTreeSet::new(),
+      timeout: request_timeout,
+      description: "echo-a-foo".to_string(),
+      jdk_home: None,
+    };
+
+    let op_name = "gimme-foo".to_string();
+
+    let mock_server = {
+      mock::execution_server::TestServer::new(
+        mock::execution_server::MockExecution::new(
+          op_name.clone(),
+          super::make_execute_request(&execute_request, empty_request_metadata())
+              .unwrap()
+              .2,
+          vec![
+            make_incomplete_operation(&op_name),
+            make_delayed_incomplete_operation(&op_name, delayed_operation_time),
+          ],
+        ),
+        None,
+      )
+    };
+
+    let cas = mock::StubCAS::builder()
+        .file(&TestData::roland())
+        .directory(&TestDirectory::containing_roland())
+        .build();
+    let command_runner = create_command_runner(mock_server.address(), &cas);
+    let mut runtime = tokio::runtime::Runtime::new().unwrap();
+
+    let successful_mock_result = FallibleExecuteProcessResult {
+      stdout: as_bytes("foo-fast"),
+      stderr: as_bytes(""),
+      exit_code: 0,
+      output_directory: EMPTY_DIGEST,
+      execution_attempts: vec![],
+    };
+
+    let run_future = command_runner.run(execute_request, WorkUnitStore::new());
+    let faster_future = Delay::new(Instant::now() + Duration::from_secs(1)).map_err(|err| format!("Error {}", err)).map({
+      let successful_mock_result = successful_mock_result.clone();
+      |_| successful_mock_result
+    });
+
+    let result = runtime.block_on(run_future.select(faster_future).map(|(result, _future)| result).map_err(|(err, _future)| err)).unwrap();
+
+    assert_eq!(result.without_execution_attempts(), successful_mock_result);
+
+    std::thread::sleep(Duration::from_secs(2));
```

Comment:
Instead of sleeping, and hoping this takes 2 seconds (which it may take longer, or it may take shorter and we'd be unnecessarily slowing down our tests), you can:
`runtime.shutdown_on_idle().wait().unwrap();`